### PR TITLE
New format-based {Encode,Decode}Key.

### PR DESCRIPTION
**Pull request by petermattis**
Added {Encode,Decode}BytesDecreasing. Added fast version of
onesComplement based on the fastXORBytes in the crypto standard
library. Changed DecodeBytes to take a result parameter to allow the
caller to avoid allocations (passing nil is fine).

### CODE REVIEW

File: util/encoding/encoding.go [hunk 1]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+					} else {
+						b = EncodeUint64Decreasing(b, args[0].(uint64))
+					}
+				}
+			case 's':
+				var arg []byte
+				if s, ok := args[0].(string); ok {
+					// Treat the string as a []byte to avoid an allocation when
+					// converting to a string. This is kosher only because the
+					// string header and slice header match on the first two
```
Comment (bdarnell) [inline util/encoding/encoding.go]: Is this property guaranteed, or could it change in other versions of go?

File: util/encoding/encoding.go [hunk 2]
```diff
@@ -398,12 +399,20 @@ func EncodeBytes(b []byte, data []byte) []byte {
 	return append(b, escape1, escapedTerm)
 }
 
+// EncodeBytesDecreasing encodes the []byte value using an
```
Comment (spencerkimball) [inline util/encoding/encoding.go]: The escape-based encoding isn't why it sorts in reverse order. Maybe reword as:

"encodes the []byte value using an escape-based encoding and then ones-complement so that it sorts in reverse order"

File: util/encoding/encoding.go [hunk 3]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+// expected and no attempt is made to perform type conversion. For
+// example, '%d' specifies an argument type of 'int64'. A panic will
+// occur if any other type is encountered.
+//
+// The verbs:
+//	%d	varint (int64) increasing
+//	%-d	varint (int64) decreasing
+//	%u	uvarint (uint64) increasing
+//	%-u	uvarint (uint64) increasing
+//	%32u	uint32 increasing
```
Comment (spencerkimball) [inline util/encoding/encoding.go]: Why are there not 32d, -32d, 64d, and -64d specifiers?

File: util/encoding/encoding.go [hunk 4]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+			panic("format mismatch")
+		}
+		i++
+	}
+	if i > 0 {
+		b = b[i:]
+	}
+
+	for i < end {
+		if format[i] != '%' {
```
Comment (bdarnell) [inline util/encoding/encoding.go]: The format string parsing should be shared between encode and decode. Pull out a function that returns a tuple (ascending, width, type).

File: util/encoding/encoding.go [hunk 5]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+				width = 32
+				i += 2
+			} else if format[i] == '6' && format[i+1] == '4' {
+				width = 64
+				i += 2
+			}
+		}
+
+		// Process verb.
+		if i < end {
```
Comment (bdarnell) [inline util/encoding/encoding.go]: It looks like this fails silently if the string ends with an unterminated specifier; this `if` should probably have an `else` that panics.

File: util/encoding/encoding.go [hunk 6]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
 		b = b[i+2:]
 	}
 }
+
+// DecodeBytesDecreasing decodes a []byte value from the input buffer
+// which was encoded using EncodeBytesDecreasing. The remainder of the
+// input buffer and the decoded []byte are returned.
+func DecodeBytesDecreasing(b []byte, r []byte) ([]byte, []byte) {
```
Comment (bdarnell) [inline util/encoding/encoding.go]: Could you eliminate this duplication by making a decodeBytesHelper that takes the escape characters as arguments?

File: util/encoding/encoding.go [hunk 7]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+		// Process width specifier (either blank, "32" or "64").
+		width := 0
+		if i+1 < end {
+			if format[i] == '3' && format[i+1] == '2' {
+				width = 32
+				i += 2
+			} else if format[i] == '6' && format[i+1] == '4' {
+				width = 64
+				i += 2
+			}
```
Comment (spencerkimball) [inline util/encoding/encoding.go]: I think you'd benefit from an additional case (in terms of helping wayward users):

    } else if format[i] >= '0' && format[i] <= '9' {
      panic("invalid integer width specifier")
    }

**Comment by bdarnell** [comment]
LGTM

File: util/encoding/encoding.go [hunk 8]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
```
Comment (spencerkimball) [inline util/encoding/encoding.go]: Ditto

**Comment by spencerkimball** [comment]
LGTM. Nice tests.

File: util/encoding/encoding.go [hunk 9]
```diff
@@ -398,12 +399,20 @@ func EncodeBytes(b []byte, data []byte) []byte {
...
+	return b
+}
+
 // DecodeBytes decodes a []byte value from the input buffer which was
 // encoded using EncodeBytes. The remainder of the input buffer and
 // the decoded []byte are returned.
-func DecodeBytes(b []byte) ([]byte, []byte) {
-	var r []byte
-
+func DecodeBytes(b []byte, r []byte) ([]byte, []byte) {
```
Comment (tschottdorf) [inline util/encoding/encoding.go]: update the comment to mention the second parameter.

File: util/encoding/encoding.go (hunk 6 above)
Comment (tschottdorf) [inline util/encoding/encoding.go]: Mention the second parameter.

File: util/encoding/encoding.go [hunk 10]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+	for i < end {
+		if format[i] != '%' {
+			panic("invalid format string: " + format)
+		}
+		i++
+
+		// Process ascending flag (either "+" or "-").
+		ascending := true
+		if i < end {
+			if format[i] == '+' {
```
Comment (tschottdorf) [inline util/encoding/encoding.go]: why not
```
if i < end {
  ascending = format[i] != '-'
  i++
}
```

File: util/encoding/encoding.go [hunk 11]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+				ascending = false
+				i++
+			}
+		}
+
+		// Process width specifier (either blank, "32" or "64").
+		width := 0
+		if i+1 < end {
+			if format[i] == '3' && format[i+1] == '2' {
+				width = 32
```
Comment (tschottdorf) [inline util/encoding/encoding.go]: doesn't seem like you're using the int value, this could as well be a bool and then you could set it directly.

File: util/encoding/encoding.go [hunk 12]
```diff
@@ -444,3 +453,255 @@ func DecodeBytes(b []byte) ([]byte, []byte) {
...
+			switch format[i] {
+			case 'd':
+				if ascending {
+					b = EncodeVarint(b, args[0].(int64))
+				} else {
+					b = EncodeVarintDecreasing(b, args[0].(int64))
+				}
+			case 'u':
+				if ascending {
+					if width == 0 {
```
Comment (tschottdorf) [inline util/encoding/encoding.go]: this case would go if  you just had width as bool.

**Comment by tschottdorf** [comment]
LGTM

File: util/encoding/encoding.go (hunk 2 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Done.

File: util/encoding/encoding.go (hunk 9 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Done.

File: util/encoding/encoding.go (hunk 6 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Mentioned the second parameter.

Comment (petermattis) [inline util/encoding/encoding.go]: @bdarnell Removed the duplication per your suggestion.

File: util/encoding/encoding.go (hunk 3 above)
Comment (petermattis) [inline util/encoding/encoding.go]: We don't have `EncodeInt{32,64}{,Decreasing}`. Doesn't seem worth adding those routines. Do you think I should add the format specifiers and cast from `int{32,64}` to `uint{32,64}`?

Comment (spencerkimball) [inline util/encoding/encoding.go]: Nah...just leave it. I was curious why the omission but until they're necessary, why bother?

File: util/encoding/encoding.go (hunk 11 above)
Comment (petermattis) [inline util/encoding/encoding.go]: I am using the int value. I distinguish between `width == {0,32,64}`.

File: util/encoding/encoding.go (hunk 7 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Done.

File: util/encoding/encoding.go (hunk 12 above)
Comment (petermattis) [inline util/encoding/encoding.go]: But I'm using `width` as an int.

File: util/encoding/encoding.go (hunk 1 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Guaranteed? I don't know. Let me see what else I can do.

File: util/encoding/encoding.go (hunk 4 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Ok. I've pulled out the parsing into a `parseVerb` function.

File: util/encoding/encoding.go (hunk 8 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Done.

File: util/encoding/encoding.go (hunk 12 above)
Comment (tschottdorf) [inline util/encoding/encoding.go]: oh, didn't see the "var" in Uvarint. Sorry.

File: util/encoding/encoding.go (hunk 1 above)
Comment (petermattis) [inline util/encoding/encoding.go]: I've added an initialization time check that this is kosher.
